File: python/day05.py
from functools import reduce


def get_id(seat):
    return reduce(lambda res, c: res * 2 + (c in 'BR'), seat, 0)


def solution():
    lines = [x.strip() for x in open('../data/05.txt', 'r').readlines()]
    ids = [get_id(line) for line in lines]

    idset = set(ids)
    seats = [id for id in range(1 << len(lines[0]))
             if id not in idset and id + 1 in idset and id - 1 in idset]

    print(f'Answer 1: {max(ids)}')
    print(f'Answer 2: {seats[0]}')

File: rust/src/day05.rs
use crate::common;
use std::collections::HashSet;

fn get_id(seat: &str) -> u32 {
  seat
    .chars()
    .fold(0, |res, c| res * 2 + "BR".contains(c) as u32)
}

pub(crate) fn solution() {
  let lines = common::read_lines(&common::data_file(5)).unwrap();
  let ids: Vec<u32> = lines.iter().map(|x| get_id(x)).collect();

  let max_id: u32 = *ids.iter().max().unwrap();
  let id_set = ids.iter().cloned().collect::<HashSet<u32>>();

  let mut seat: u32 = 0;
  for i in 0..=(1 << lines[0].len()) {
    let id = i as u32;
    if !id_set.contains(&id) && id_set.contains(&(id + 1)) && id_set.contains(&(id - 1)) {
      seat = id;
      break;
    }
  }
  println!("Max ID: {}, Seat: {}", max_id, seat);
}
